File: 169-majority-element/169.js
// solution 1
var majorityElement = function(nums) {
  let count = new Map();
  for (let n of nums) {
    if (!count.has(n)) {
      count.set(n, 0);
    }
    count.set(n, count.get(n) + 1);

    if (count.get(n) > nums.length / 2) {
      return n;
    }
  }
};

//solution 2
var majorityElement = function(nums) {
  nums.sort();
  return nums[Math.floor(nums.length / 2)];
};

// solution 3
var majorityElement = function(nums) {
  let maj;
  let count = 0;

  for (let n of nums) {
    if (count === 0) {
      maj = n;
    }
    if (n === maj) {
      count++;
    } else {
      count--;
    }
  }
  return maj;
};

File: array/169/169.py
# solution 1
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        count = {}
        for n in nums:
            if n not in count:
                count[n] = 0
            count[n] += 1
            if count[n] > len(nums)/2:
                return n


# solution 2
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        nums.sort()
        # double slash is for integer division
        return nums[len(nums)//2]


# solution 3
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        maj = None
        count = 0

        for n in nums:
            if count == 0:
                maj = n
            count += (1 if n == maj else -1)
        return maj
